Thread-Safety issues during shutdown

Race detection reports issues during tunnel shutdown.'

```
==================
WARNING: DATA RACE
Write at 0x00c0006fe030 by goroutine 74:
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.TestTunnelServiceInitialization()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/client-go/testing/fake.go:102 +0xf49
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:827 +0x162

Previous read at 0x00c0006fe030 by goroutine 92:
  [failed to restore the stack]

Goroutine 74 (running) created at:
  testing.(*T).Run()
      /usr/local/opt/go/libexec/src/testing/testing.go:878 +0x650
  testing.runTests.func1()
      /usr/local/opt/go/libexec/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/opt/go/libexec/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/opt/go/libexec/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:64 +0x221

Goroutine 92 (running) created at:
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).Run()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:330 +0x420
==================
==================
WARNING: DATA RACE
Read at 0x00c000026840 by goroutine 74:
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.TestTunnelServiceInitialization()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller_test.go:445 +0x1309
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:827 +0x162

Previous write at 0x00c000026840 by goroutine 92:
  runtime.mapassign_faststr()
      /usr/local/opt/go/libexec/src/runtime/map_faststr.go:190 +0x0
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).setTunnel()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:83 +0xbe
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).createTunnel()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:625 +0x750
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).processIngress()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:420 +0xca4
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).processNextIngress()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:351 +0x11f
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).runIngressWorker()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:339 +0x38
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).runIngressWorker-fm()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:330 +0x41
  github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x61
  github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xce
  github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x5a

Goroutine 74 (running) created at:
  testing.(*T).Run()
      /usr/local/opt/go/libexec/src/testing/testing.go:878 +0x650
  testing.runTests.func1()
      /usr/local/opt/go/libexec/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/opt/go/libexec/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/opt/go/libexec/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:64 +0x221

Goroutine 92 (running) created at:
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).Run()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:330 +0x420
==================
I0917 07:24:58.705336   18311 controller.go:334] Stopping ArgoController 
--- FAIL: TestTunnelServiceInitialization (6.01s)
    testing.go:771: race detected during execution of test
==================
WARNING: DATA RACE
Read at 0x00c0001ab950 by goroutine 94:
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.TestTunnelServicesTwoNS()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller_test.go:535 +0x1592
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:827 +0x162

Previous write at 0x00c0001ab950 by goroutine 110:
  runtime.mapassign_faststr()
      /usr/local/opt/go/libexec/src/runtime/map_faststr.go:190 +0x0
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).setTunnel()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:83 +0xbe
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).createTunnel()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:625 +0x750
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).processIngress()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:420 +0xca4
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).processNextIngress()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:351 +0x11f
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).runIngressWorker()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:339 +0x38
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).runIngressWorker-fm()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:330 +0x41
  github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x61
  github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xce
  github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x5a

Goroutine 94 (running) created at:
  testing.(*T).Run()
      /usr/local/opt/go/libexec/src/testing/testing.go:878 +0x650
  testing.runTests.func1()
      /usr/local/opt/go/libexec/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/opt/go/libexec/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/opt/go/libexec/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/opt/go/libexec/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:64 +0x221

Goroutine 110 (running) created at:
  github.com/cloudflare/cloudflare-ingress-controller/internal/controller.(*ArgoController).Run()
      /Users/malberts/go/src/github.com/cloudflare/cloudflare-ingress-controller/internal/controller/controller.go:330 +0x420
==================
I0917 07:25:03.816015   18311 watch.go:109] Stopping fake watcher.
--- FAIL: TestTunnelServicesTwoNS (5.11s)
    testing.go:771: race detected during execution of test
```

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -406,9 +406,20 @@
   version = "v1.0.1"
 
 [[projects]]
+  digest = "1:711eebe744c0151a9d09af2315f0bb729b2ec7637ef4c410fa90a18ef74b65b6"
+  name = "github.com/stretchr/objx"
+  packages = ["."]
+  pruneopts = ""
+  revision = "477a77ecc69700c7cdeb1fa9e129548e1c1c393c"
+  version = "v0.1.1"
+
+[[projects]]
   digest = "1:a30066593578732a356dc7e5d7f78d69184ca65aeeff5939241a3ab10559bb06"
   name = "github.com/stretchr/testify"
-  packages = ["assert"]
+  packages = [
+    "assert",
+    "mock",
+  ]
   pruneopts = ""
   revision = "12b6f73e6084dad08a7c6e575284b177ecafbc71"
   version = "v1.2.1"
@@ -716,8 +727,10 @@
     "github.com/pkg/errors",
     "github.com/sirupsen/logrus",
     "github.com/stretchr/testify/assert",
+    "github.com/stretchr/testify/mock",
     "k8s.io/api/core/v1",
     "k8s.io/api/extensions/v1beta1",
+    "k8s.io/apimachinery/pkg/api/meta",
     "k8s.io/apimachinery/pkg/apis/meta/v1",
     "k8s.io/apimachinery/pkg/labels",
     "k8s.io/apimachinery/pkg/runtime",

--- a/Makefile
+++ b/Makefile
@@ -121,7 +121,7 @@ push: container
 staticcheck:
 	@echo static checking code for issues
 	@go get honnef.co/go/tools/cmd/staticcheck
-	staticcheck $(SRCS)
+	@staticcheck $(SRCS)
 
 .PHONY: test
 test: install
@@ -130,13 +130,14 @@ test: install
 
 .PHONY: test-race
 test-race: | test
+	@echo testing code for races
 	@go test -race ./...
 
 .PHONY: unused
 unused:
 	@echo checking code for unused definitions
 	@go get honnef.co/go/tools/cmd/unused
-	unused -exported $(SRCS)
+	@unused -exported $(SRCS)
 
 .PHONY: version
 version:

--- a/internal/controller/controller.go
+++ b/internal/controller/controller.go
@@ -3,7 +3,6 @@ package controller
 import (
 	"fmt"
 	"strings"
-	"sync"
 	"time"
 
 	"github.com/cloudflare/cloudflare-ingress-controller/internal/tunnel"
@@ -11,6 +10,7 @@ import (
 	"github.com/pkg/errors"
 	"k8s.io/api/core/v1"
 	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
 	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime"
@@ -42,16 +42,13 @@ type ArgoController struct {
 	endpointsLister   lister_v1.EndpointsLister
 	endpointsInformer cache.Controller
 
-	// map of key to tunnels
-	// where key concatenates the namespace, ingressname and servicename
-	mux     sync.Mutex
-	tunnels map[string]tunnel.Tunnel
+	// tunnel registry, key = (namespace, ingressname and servicename)
+	tunnels tunnel.Registry
 }
 
 func NewArgoController(client kubernetes.Interface, options ...Option) *ArgoController {
 	opts := collectOptions(options)
 	informer, indexer, queue := createIngressInformer(client, opts.ingressClass)
-	tunnels := make(map[string]tunnel.Tunnel, 0)
 
 	argo := &ArgoController{
 		client:        client,
@@ -61,9 +58,6 @@ func NewArgoController(client kubernetes.Interface, options ...Option) *ArgoCont
 		ingressInformer:  informer,
 		ingressWorkqueue: queue,
 		ingressLister:    lister_v1beta1.NewIngressLister(indexer),
-
-		mux:     sync.Mutex{},
-		tunnels: tunnels,
 	}
 	argo.configureServiceInformer()
 	argo.configureEndpointInformer()
@@ -74,27 +68,14 @@ func NewArgoController(client kubernetes.Interface, options ...Option) *ArgoCont
 	return argo
 }
 
-func (argo *ArgoController) getTunnel(key string) tunnel.Tunnel {
-	argo.mux.Lock()
-	defer argo.mux.Unlock()
-	return argo.tunnels[key]
-}
-
-func (argo *ArgoController) setTunnel(key string, t tunnel.Tunnel) {
-	argo.mux.Lock()
-	defer argo.mux.Unlock()
-	argo.tunnels[key] = t
-}
-
-func (argo *ArgoController) getTunnelsForService(namespace, serviceName string) []string {
-	argo.mux.Lock()
-	defer argo.mux.Unlock()
+func (argo *ArgoController) getTunnelsForService(namespace, name string) []string {
 	var keys []string
-	for key, t := range argo.tunnels {
-		if serviceName == t.Config().ServiceName && namespace == t.Config().ServiceNamespace {
-			keys = append(keys, key)
+	argo.tunnels.Range(func(k string, t tunnel.Tunnel) bool {
+		if name == t.Config().ServiceName && namespace == t.Config().ServiceNamespace {
+			keys = append(keys, k)
 		}
-	}
+		return true
+	})
 	return keys
 }
 
@@ -230,14 +211,23 @@ func (argo *ArgoController) configureServiceInformer() {
 }
 
 // is this service one of the ones we have a tunnel for?
-func (argo *ArgoController) isWatchedService(service *v1.Service) bool {
-	for _, tunnel := range argo.tunnels {
-		if service.ObjectMeta.Name == tunnel.Config().ServiceName && service.ObjectMeta.Namespace == tunnel.Config().ServiceNamespace {
-			glog.V(5).Infof("Watching service %s/%s", service.ObjectMeta.Namespace, service.ObjectMeta.Name)
-			return true
-		}
+func (argo *ArgoController) isWatchedService(service *v1.Service) (ok bool) {
+	svcMeta, err := meta.Accessor(service)
+	if err != nil {
+		return
 	}
-	return false
+
+	name, ns := svcMeta.GetName(), svcMeta.GetNamespace()
+	argo.tunnels.Range(func(k string, t tunnel.Tunnel) bool {
+		if name == t.Config().ServiceName && ns == t.Config().ServiceNamespace {
+			glog.V(5).Infof("Watching service %s/%s", ns, name)
+			// set outer return and trigger stop condition
+			ok = true
+			return false
+		}
+		return true
+	})
+	return
 }
 
 func (argo *ArgoController) configureEndpointInformer() {
@@ -290,19 +280,23 @@ func (argo *ArgoController) configureEndpointInformer() {
 }
 
 // is this endpoint interesting?
-func (argo *ArgoController) isWatchedEndpoint(ep *v1.Endpoints) bool {
-
-	// XXX fix this synchronization
-	argo.mux.Lock()
-	defer argo.mux.Unlock()
-
-	for _, tunnel := range argo.tunnels {
-		if ep.ObjectMeta.Name == tunnel.Config().ServiceName {
-			glog.V(5).Infof("Watching endpoint %s/%s", ep.ObjectMeta.Namespace, ep.ObjectMeta.Name)
-			return true
-		}
+func (argo *ArgoController) isWatchedEndpoint(ep *v1.Endpoints) (ok bool) {
+	epMeta, err := meta.Accessor(ep)
+	if err != nil {
+		return
 	}
-	return false
+
+	name, ns := epMeta.GetName(), epMeta.GetNamespace()
+	argo.tunnels.Range(func(k string, t tunnel.Tunnel) bool {
+		if name == t.Config().ServiceName {
+			glog.V(5).Infof("Watching endpoint %s/%s", ns, name)
+			// set outer return and trigger stop condition
+			ok = true
+			return false
+		}
+		return true
+	})
+	return
 }
 
 func (argo *ArgoController) Run(stopCh <-chan struct{}) {
@@ -409,8 +403,8 @@ func (argo *ArgoController) processIngress(queueKey string) error {
 		}
 		key := constructIngressKey(ingress)
 
-		tunnel := argo.getTunnel(key)
-		if tunnel != nil {
+		tunnel, ok := argo.tunnels.Load(key)
+		if ok {
 			glog.V(5).Infof("Tunnel \"%s\" (%s) already exists", key, tunnel.Config().ExternalHostname)
 			return nil
 		}
@@ -621,7 +615,7 @@ func (argo *ArgoController) createTunnel(key string, ingress *v1beta1.Ingress) e
 	if err != nil {
 		return err
 	}
-	argo.setTunnel(key, tunnel)
+	argo.tunnels.Store(key, tunnel)
 	glog.V(5).Infof("created tunnel for ingress %s,  %s", ingressName, key)
 
 	return argo.evaluateTunnelStatus(key)
@@ -634,8 +628,8 @@ func (argo *ArgoController) updateTunnel(key string, ingress *v1beta1.Ingress) e
 		argo.removeTunnel(key)
 		return err
 	}
-	t := argo.getTunnel(key)
-	if t == nil {
+	t, ok := argo.tunnels.Load(key)
+	if !ok {
 		return argo.createTunnel(key, ingress)
 	}
 
@@ -656,9 +650,8 @@ func (argo *ArgoController) updateTunnel(key string, ingress *v1beta1.Ingress) e
 // starts or stops the tunnel depending on the existence of
 // the associated service and endpoints
 func (argo *ArgoController) evaluateTunnelStatus(key string) error {
-
-	t := argo.getTunnel(key)
-	if t == nil {
+	t, ok := argo.tunnels.Load(key)
+	if !ok {
 		return fmt.Errorf("Tunnel not found for key %s", key)
 	}
 
@@ -670,8 +663,8 @@ func (argo *ArgoController) evaluateTunnelStatus(key string) error {
 		glog.V(2).Infof("Service %s not found for tunnel", key)
 		return argo.stopTunnel(t)
 	}
-	endpoints, err := argo.endpointsLister.Endpoints(namespace).Get(serviceName)
 
+	endpoints, err := argo.endpointsLister.Endpoints(namespace).Get(serviceName)
 	if err != nil || endpoints == nil {
 		glog.V(2).Infof("Endpoints not found for tunnel %s", key)
 		return argo.stopTunnel(t)
@@ -764,25 +757,28 @@ func (argo *ArgoController) stopTunnel(t tunnel.Tunnel) error {
 
 func (argo *ArgoController) removeTunnel(key string) error {
 	glog.V(5).Infof("Removing tunnel %s", key)
-	t := argo.getTunnel(key)
-	if t == nil {
+	t, ok := argo.tunnels.LoadAndDelete(key)
+	if !ok {
 		return fmt.Errorf("Tunnel not found for key %s", key)
 	}
-	err := argo.stopTunnel(t)
-	argo.mux.Lock()
-	delete(argo.tunnels, key)
-	argo.mux.Unlock()
-	return err
+	// Issue: if stopping the tunnel errors, the reference to the object
+	// will be lost; but the tunnel may not have been detached and cleaned.
+	// (the issue is historic and being preserved)
+	return argo.stopTunnel(t)
 }
 
 func (argo *ArgoController) tearDown() error {
 	glog.V(5).Infof("Tearing down tunnels")
-
-	for _, t := range argo.tunnels {
-		t.TearDown()
-	}
-	argo.mux.Lock()
-	argo.tunnels = make(map[string]tunnel.Tunnel)
-	argo.mux.Unlock()
+	var wg wait.Group
+	argo.tunnels.Filter(func(k string, t tunnel.Tunnel) bool {
+		wg.Start(func() {
+			// Issue: use of teardown vs stop is suspect.
+			if err := t.TearDown(); err != nil {
+				glog.V(2).Infof("Error halting tunnel, %v", err)
+			}
+		})
+		return true
+	})
+	wg.Wait()
 	return nil
 }

--- a/internal/controller/controller_test.go
+++ b/internal/controller/controller_test.go
@@ -14,6 +14,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/kubernetes/fake"
 	ktesting "k8s.io/client-go/testing"
@@ -44,7 +45,7 @@ func checkServiceKey(t *testing.T, check serviceKeyCheck) {
 }
 
 func TestServiceKey(t *testing.T) {
-
+	t.Parallel()
 	checks := []serviceKeyCheck{
 		{
 			"add:default/nginx",
@@ -104,7 +105,7 @@ func checkIngressKey(t *testing.T, check ingressKeyCheck) {
 }
 
 func TestIngressKey(t *testing.T) {
-
+	t.Parallel()
 	ingressSpecNginx := v1beta1.IngressSpec{
 		Rules: []v1beta1.IngressRule{
 			{
@@ -174,6 +175,7 @@ func TestIngressKey(t *testing.T) {
 }
 
 func TestSimple(t *testing.T) {
+	t.Parallel()
 	fakeClient := &fake.Clientset{}
 	nothingPod := v1.Pod{}
 
@@ -193,6 +195,7 @@ func TestSimple(t *testing.T) {
 }
 
 func TestAction(t *testing.T) {
+	t.Parallel()
 	serviceNamespace := "acme"
 	fakeClient := &fake.Clientset{}
 	actualPod := v1.Pod{
@@ -224,6 +227,7 @@ func TestAction(t *testing.T) {
 }
 
 func TestNewArgoController(t *testing.T) {
+	t.Parallel()
 	controllerNamespace := "cloudflare" // "cloudflare"
 	fakeClient := &fake.Clientset{}
 
@@ -234,12 +238,14 @@ func TestNewArgoController(t *testing.T) {
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 	go wc.Run(stopCh)
-	// wait for cache sync
-	time.Sleep(time.Second)
+
+	cache.WaitForCacheSync(stopCh,
+		wc.ingressInformer.HasSynced,
+		wc.serviceInformer.HasSynced,
+	)
 
 	assert.Equal(t, wc.options.secretNamespace, controllerNamespace)
-	assert.NotNil(t, wc.tunnels)
-	assert.Equal(t, 0, len(wc.tunnels))
+	assert.Equal(t, 0, wc.tunnels.Len())
 }
 
 type tunnelItems struct {
@@ -309,6 +315,7 @@ func getTunnelItems(namespace string) tunnelItems {
 }
 
 func TestControllerLookups(t *testing.T) {
+	t.Parallel()
 	fakeClient := &fake.Clientset{}
 
 	serviceNamespace := "acme"
@@ -328,6 +335,7 @@ func TestControllerLookups(t *testing.T) {
 }
 
 func TestTunnelInitialization(t *testing.T) {
+	t.Parallel()
 	fakeClient := &fake.Clientset{}
 
 	serviceNamespace := "acme"
@@ -364,19 +372,22 @@ func TestTunnelInitialization(t *testing.T) {
 	defer close(stopCh)
 	go wc.Run(stopCh)
 
-	// wait for cache sync
-	time.Sleep(time.Second)
+	cache.WaitForCacheSync(stopCh,
+		wc.ingressInformer.HasSynced,
+		wc.serviceInformer.HasSynced,
+	)
+
+	wait.Poll(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
+		done = wc.tunnels.Len() > 0
+		return
+	})
 
 	assert.Equal(t, wc.options.secretNamespace, controllerNamespace)
-	if wc.tunnels == nil {
-		t.Fatal("failing, tunnels is nil")
-	}
-	assert.NotNil(t, wc.tunnels)
-	assert.Equal(t, 1, len(wc.tunnels))
+	assert.Equal(t, 1, wc.tunnels.Len())
 
 	key := constructIngressKey(&items.Ingress)
-	fooTunnel := wc.getTunnel(key)
-	if fooTunnel == nil {
+	fooTunnel, ok := wc.tunnels.Load(key)
+	if !ok {
 		t.Fatalf("failing, tunnel is nil for %s", key)
 	}
 	assert.False(t, fooTunnel.Active())
@@ -387,6 +398,7 @@ func TestTunnelInitialization(t *testing.T) {
 }
 
 func TestTunnelServiceInitialization(t *testing.T) {
+	t.Parallel()
 	fakeClient := &fake.Clientset{}
 
 	controllerNamespace := "cloudflare"
@@ -420,8 +432,10 @@ func TestTunnelServiceInitialization(t *testing.T) {
 	defer close(stopCh)
 	go wc.Run(stopCh)
 
-	// wait for cache sync
-	time.Sleep(time.Second)
+	cache.WaitForCacheSync(stopCh,
+		wc.ingressInformer.HasSynced,
+		wc.serviceInformer.HasSynced,
+	)
 
 	// add the service now
 	fakeClient.Fake.AddReactor("list", "services", func(action ktesting.Action) (bool, runtime.Object, error) {
@@ -434,17 +448,17 @@ func TestTunnelServiceInitialization(t *testing.T) {
 	getServiceAction := ktesting.NewCreateAction(serviceResource, serviceNamespace, &items.Service)
 	_, _ = fakeClient.Invokes(getServiceAction, &v1.Service{})
 
-	time.Sleep(5 * time.Second)
+	wait.Poll(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
+		done = wc.tunnels.Len() > 0
+		return
+	})
 
 	assert.Equal(t, wc.options.secretNamespace, controllerNamespace)
-	if wc.tunnels == nil {
-		t.Fatal("failing, tunnels is nil")
-	}
-	assert.Equal(t, 1, len(wc.tunnels))
+	assert.Equal(t, 1, wc.tunnels.Len())
 
 	key := constructIngressKey(&items.Ingress)
-	fooTunnel := wc.getTunnel(key)
-	if fooTunnel == nil {
+	fooTunnel, ok := wc.tunnels.Load(key)
+	if !ok {
 		t.Fatalf("failing, tunnel is nil for %s", key)
 	}
 	assert.False(t, fooTunnel.Active())
@@ -455,6 +469,7 @@ func TestTunnelServiceInitialization(t *testing.T) {
 }
 
 func TestTunnelServicesTwoNS(t *testing.T) {
+	t.Parallel()
 	fakeClient := &fake.Clientset{}
 
 	controllerNamespace := "cloudflare"
@@ -522,19 +537,18 @@ func TestTunnelServicesTwoNS(t *testing.T) {
 	getServiceAction := ktesting.NewCreateAction(serviceResource, items[0].Ingress.GetNamespace(), &items[0].Service)
 	_, _ = fakeClient.Invokes(getServiceAction, &v1.Service{})
 
-	// XXX - fix explicit sleep
-	time.Sleep(5 * time.Second)
+	wait.Poll(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
+		done = wc.tunnels.Len() > 0
+		return
+	})
 
 	assert.Equal(t, wc.options.secretNamespace, controllerNamespace)
-	if wc.tunnels == nil {
-		t.Fatal("failing, tunnels is nil")
-	}
-	assert.Equal(t, 2, len(wc.tunnels))
+	assert.Equal(t, 2, wc.tunnels.Len())
 
 	for _, item := range items {
 		key := constructIngressKey(&item.Ingress)
-		tunnel := wc.getTunnel(key)
-		if tunnel == nil {
+		tunnel, ok := wc.tunnels.Load(key)
+		if !ok {
 			t.Fatalf("failing, tunnel is nil for %s", key)
 		}
 		assert.False(t, tunnel.Active())

--- a/internal/tunnel/argo.go
+++ b/internal/tunnel/argo.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"net"
 	"net/http"
+	"sync"
 	"time"
 
 	"github.com/cloudflare/cloudflare-ingress-controller/internal/cloudflare"
@@ -22,10 +23,10 @@ type ArgoTunnelManager struct {
 	tunnelConfig *origin.TunnelConfig
 	errCh        chan error
 	stopCh       chan struct{}
+	mu           sync.RWMutex
 }
 
 func newHttpTransport() *http.Transport {
-
 	tlsConfig := &tls.Config{}
 
 	httpTransport := &http.Transport{
@@ -104,22 +105,28 @@ func (mgr *ArgoTunnelManager) Config() Config {
 }
 
 func (mgr *ArgoTunnelManager) Active() bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
 	return mgr.stopCh != nil
 }
 
 func (mgr *ArgoTunnelManager) Start(serviceURL string) error {
-
 	if serviceURL == "" {
 		return fmt.Errorf("Cannot start tunnel for %s with empty url", mgr.Config().ServiceName)
+	} else if mgr.stopCh != nil {
+		return nil
 	}
-	mgr.tunnelConfig.OriginUrl = serviceURL
 
-	placeHolderOnlyConnectedSignal := make(chan struct{})
-	mgr.stopCh = make(chan struct{})
-
-	go func() {
-		mgr.errCh <- origin.StartTunnelDaemon(mgr.tunnelConfig, mgr.stopCh, placeHolderOnlyConnectedSignal)
-	}()
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	if mgr.stopCh == nil {
+		mgr.tunnelConfig.OriginUrl = serviceURL
+		dummyChan := make(chan struct{})
+		mgr.stopCh = make(chan struct{})
+		go func() {
+			mgr.errCh <- origin.StartTunnelDaemon(mgr.tunnelConfig, mgr.stopCh, dummyChan)
+		}()
+	}
 	return nil
 }
 
@@ -127,10 +134,15 @@ func (mgr *ArgoTunnelManager) Stop() error {
 	if mgr.stopCh == nil {
 		return fmt.Errorf("tunnel %s already stopped", mgr.id)
 	}
-	close(mgr.stopCh)
-	mgr.tunnelConfig.OriginUrl = ""
-	mgr.stopCh = nil
-	<-mgr.errCh // muxerShutdownError is not an error
+
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	if mgr.stopCh != nil {
+		close(mgr.stopCh)
+		mgr.tunnelConfig.OriginUrl = ""
+		mgr.stopCh = nil
+		<-mgr.errCh // muxerShutdownError is not an error
+	}
 	return nil
 }
 

--- a/internal/tunnel/registry.go
+++ b/internal/tunnel/registry.go
@@ -1,0 +1,78 @@
+package tunnel
+
+import (
+	"sync"
+)
+
+// Registry key/tunnel map protected by a RWMutex
+// The registry is a temporary mechanism used to enforce thread-safety.
+// It is likely to replaced/removed with an update to the informers.
+type Registry struct {
+	mu  sync.RWMutex
+	obj map[string]Tunnel
+}
+
+// Load recovers a tunnel by key
+func (r *Registry) Load(key string) (val Tunnel, ok bool) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	val, ok = r.obj[key]
+	return
+}
+
+// Store registers a tunnel by key
+func (r *Registry) Store(key string, val Tunnel) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.obj == nil {
+		r.obj = make(map[string]Tunnel)
+	}
+	r.obj[key] = val
+}
+
+// Delete removes a tunnel by key
+func (r *Registry) Delete(key string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	delete(r.obj, key)
+	return
+}
+
+// LoadAndDelete deletes the tunnel by key and returns the stored value
+func (r *Registry) LoadAndDelete(key string) (val Tunnel, ok bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if val, ok = r.obj[key]; ok {
+		delete(r.obj, key)
+	}
+	return
+}
+
+// Range processes the function on all tunnels (false terminates iteration).
+func (r *Registry) Range(f func(key string, val Tunnel) bool) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	for k, v := range r.obj {
+		if !f(k, v) {
+			break
+		}
+	}
+}
+
+// Filter processes the function on all tunnels (true deletes the tunnel).
+func (r *Registry) Filter(f func(key string, val Tunnel) bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	for k, v := range r.obj {
+		if f(k, v) {
+			delete(r.obj, k)
+		}
+	}
+}
+
+// Len returns the number of items in the registry
+func (r *Registry) Len() int {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return len(r.obj)
+}

--- a/internal/tunnel/registry_test.go
+++ b/internal/tunnel/registry_test.go
@@ -1,0 +1,472 @@
+package tunnel
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// The tunnel interface does not provide a unique identifier for the
+// tunnel.  In the tests below, tunnel is mocked to provide a dummy
+// implemention where config can be used to setup and recover an
+// identifying marker (ServiceName) for testing.
+
+func TestLoad(t *testing.T) {
+	t.Parallel()
+	for name, test := range map[string]struct {
+		obj map[string]Tunnel
+		key string
+		tag string
+		ok  bool
+	}{
+		"nil-map": {
+			obj: nil,
+			key: "test",
+			tag: "",
+			ok:  false,
+		},
+		"not-found": {
+			obj: map[string]Tunnel{
+				"unit": nil,
+			},
+			key: "test",
+			tag: "",
+			ok:  false,
+		},
+		"load": {
+			obj: map[string]Tunnel{
+				"test": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test",
+					})
+					return t
+				}(),
+			},
+			key: "test",
+			tag: "test",
+			ok:  true,
+		},
+	} {
+		r := Registry{
+			obj: test.obj,
+		}
+
+		_, ok := r.Load(test.key)
+		tag := func() (tag string) {
+			if r.obj != nil {
+				if t, ok := r.obj[test.key]; ok {
+					tag = t.Config().ServiceName
+				}
+			}
+			return
+		}()
+		assert.Equalf(t, test.ok, ok, "test '%s' ok mismatch", name)
+		assert.Equalf(t, test.tag, tag, "test '%s' tag mismatch", name)
+	}
+}
+
+func TestStore(t *testing.T) {
+	t.Parallel()
+	for name, test := range map[string]struct {
+		obj map[string]Tunnel
+		key string
+		val Tunnel
+		len int
+	}{
+		"nil-map": {
+			obj: nil,
+			key: "test",
+			val: func() Tunnel {
+				t := &mockTunnel{}
+				t.On("Config").Return(Config{
+					ServiceName: "test",
+				})
+				return t
+			}(),
+			len: 1,
+		},
+		"append": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+			},
+			key: "test-b",
+			val: func() Tunnel {
+				t := &mockTunnel{}
+				t.On("Config").Return(Config{
+					ServiceName: "test-b",
+				})
+				return t
+			}(),
+			len: 2,
+		},
+		"overwrite": {
+			obj: map[string]Tunnel{
+				"test": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+			},
+			key: "test",
+			val: func() Tunnel {
+				t := &mockTunnel{}
+				t.On("Config").Return(Config{
+					ServiceName: "test-b",
+				})
+				return t
+			}(),
+			len: 1,
+		},
+	} {
+		r := Registry{
+			obj: test.obj,
+		}
+
+		r.Store(test.key, test.val)
+		val := func() (t Tunnel) {
+			if r.obj != nil {
+				t, _ = r.obj[test.key]
+			}
+			return
+		}()
+		assert.Equalf(t, test.val, val, "test '%s' store mismatch", name)
+		assert.Equalf(t, test.len, len(r.obj), "test '%s' length mismatch", name)
+	}
+}
+
+func TestDelete(t *testing.T) {
+	t.Parallel()
+	for name, test := range map[string]struct {
+		obj map[string]Tunnel
+		key string
+		len int
+	}{
+		"nil-map": {
+			obj: nil,
+			key: "test",
+			len: 0,
+		},
+		"not-found": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+			},
+			key: "test",
+			len: 1,
+		},
+		"remove": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+			},
+			key: "test-a",
+			len: 1,
+		},
+	} {
+		r := Registry{
+			obj: test.obj,
+		}
+
+		r.Delete(test.key)
+		out := func() (t Tunnel) {
+			if r.obj != nil {
+				t, _ = r.obj[test.key]
+			}
+			return
+		}()
+		assert.Emptyf(t, out, "test '%s' delete found non-nil tunnel", name)
+		assert.Equalf(t, test.len, len(r.obj), "test '%s' length mismatch", name)
+	}
+}
+
+func TestLoadAndDelete(t *testing.T) {
+	t.Parallel()
+	for name, test := range map[string]struct {
+		obj map[string]Tunnel
+		key string
+		tag string
+		ok  bool
+		len int
+	}{
+		"nil-map": {
+			obj: nil,
+			key: "test",
+			ok:  false,
+			tag: "",
+			len: 0,
+		},
+		"not-found": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+			},
+			key: "test",
+			ok:  false,
+			tag: "",
+			len: 1,
+		},
+		"load-and-remove": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+			},
+			key: "test-b",
+			ok:  true,
+			tag: "test-b",
+			len: 1,
+		},
+	} {
+		r := Registry{
+			obj: test.obj,
+		}
+
+		val, ok := r.LoadAndDelete(test.key)
+		tag := func() (tag string) {
+			if val != nil {
+				tag = val.Config().ServiceName
+			}
+			return
+		}()
+		out := func() (t Tunnel) {
+			if r.obj != nil {
+				t, _ = r.obj[test.key]
+			}
+			return
+		}()
+		assert.Emptyf(t, out, "test '%s' delete found non-nil tunnel", name)
+		assert.Equalf(t, test.len, len(r.obj), "test '%s' length mismatch", name)
+		assert.Equalf(t, test.ok, ok, "test '%s' ok mismatch", name)
+		assert.Equalf(t, test.tag, tag, "test '%s' tag mismatch", name)
+	}
+}
+
+func TestRange(t *testing.T) {
+	t.Parallel()
+	for name, test := range map[string]struct {
+		obj map[string]Tunnel
+		key string
+		len int
+	}{
+		"nil-map": {
+			obj: nil,
+			key: "test-a",
+			len: 0,
+		},
+		"visit-none": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+				"test-c": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-c",
+					})
+					return t
+				}(),
+			},
+			key: "test-",
+			len: 1,
+		},
+		"visit-all": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+				"test-c": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-c",
+					})
+					return t
+				}(),
+			},
+			key: "test-z",
+			len: 3,
+		},
+	} {
+		r := Registry{
+			obj: test.obj,
+		}
+
+		len := 0
+		r.Range(func(k string, t Tunnel) bool {
+			len++
+			return !strings.Contains(k, test.key)
+		})
+
+		assert.Equalf(t, test.len, len, "test '%s' length mismatch", name)
+	}
+}
+
+func TestFilter(t *testing.T) {
+	t.Parallel()
+	for name, test := range map[string]struct {
+		obj map[string]Tunnel
+		key string
+		len int
+	}{
+		"nil-map": {
+			obj: nil,
+			key: "test-a",
+			len: 0,
+		},
+		"filter-none": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+				"test-c": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-c",
+					})
+					return t
+				}(),
+			},
+			key: "test-z",
+			len: 3,
+		},
+		"filter-one": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+				"test-c": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-c",
+					})
+					return t
+				}(),
+			},
+			key: "test-b",
+			len: 2,
+		},
+		"filter-all": {
+			obj: map[string]Tunnel{
+				"test-a": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-a",
+					})
+					return t
+				}(),
+				"test-b": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-b",
+					})
+					return t
+				}(),
+				"test-c": func() Tunnel {
+					t := &mockTunnel{}
+					t.On("Config").Return(Config{
+						ServiceName: "test-c",
+					})
+					return t
+				}(),
+			},
+			key: "test-",
+			len: 0,
+		},
+	} {
+		r := Registry{
+			obj: test.obj,
+		}
+
+		r.Filter(func(k string, t Tunnel) bool {
+			return strings.Contains(k, test.key)
+		})
+
+		assert.Equalf(t, test.len, len(r.obj), "test '%s' length mismatch", name)
+	}
+}

--- a/internal/tunnel/tunnel_test.go
+++ b/internal/tunnel/tunnel_test.go
@@ -1,0 +1,39 @@
+package tunnel
+
+import (
+	"github.com/stretchr/testify/mock"
+)
+
+type mockTunnel struct {
+	mock.Mock
+}
+
+func (t *mockTunnel) Config() Config {
+	args := t.Called()
+	return args.Get(0).(Config)
+}
+
+func (t *mockTunnel) Start(serviceURL string) error {
+	args := t.Called(serviceURL)
+	return args.Error(0)
+}
+
+func (t *mockTunnel) Stop() error {
+	args := t.Called()
+	return args.Error(0)
+}
+
+func (t *mockTunnel) Active() bool {
+	args := t.Called()
+	return args.Get(0).(bool)
+}
+
+func (t *mockTunnel) TearDown() error {
+	args := t.Called()
+	return args.Error(0)
+}
+
+func (t *mockTunnel) CheckStatus() error {
+	args := t.Called()
+	return args.Error(0)
+}

